[BUG] Possible regression in Azure.Resource.AllowedRegions

### Existing rule

Azure.Resource.AllowedRegions

### Description of the issue

The rule fails even though we haven't configured allowed locations with `Azure_AllowedRegions` or the new `AZURE_RESOURCE_ALLOWED_LOCATIONS` configuration settings.

This started happening when PSRule installed PSRule.Rules.Azure v1.30, where it was using v1.29 before.

### Error messages

```
 -> 66c6950b43a62 : Microsoft.Web/sites [9/10]

    [PASS] Azure.Resource.UseTags (AZR-000166)
    [FAIL] Azure.Resource.AllowedRegions (AZR-000167)
    | Template: templates/function.bicep:193:5

    | RECOMMEND:
    | Consider deploying resources to allowed regions to align with your
    | organizational requirements. Also consider using Azure Policy to enforce
    | allowed regions.

    | REASON:
    | - The location 'eastus' is not in the allowed set of resource locations.

    | HELP:
    | - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Resource.AllowedRegions/

##[error]AZR-000167: 66c6950b43a62 failed Azure.Resource.AllowedRegions. Resources should be deployed to allowed regions.
```

### Reproduction

Here's our config file:

```yaml
binding:
  preferTargetInfo: true
  targetType:
    - type
    - resourceType

# Require minimum versions of modules.
requires:
  PSRule: '@pre >=2.9.0'
  PSRule.Rules.Azure: '@pre >=1.28.0'

# Use PSRule for Azure.
include:
  module:
    - PSRule.Rules.Azure

output:
  culture:
    - 'en-AU'

execution:
  unprocessedObject: Ignore

configuration:
  # Enable Bicep CLI checks.
  AZURE_BICEP_CHECK_TOOL: true

  # Enable automatic expansion of Azure parameter files.
  AZURE_PARAMETER_FILE_EXPANSION: true

  # Enable automatic expansion of Azure Bicep source files.
  AZURE_BICEP_FILE_EXPANSION: true

  # Configures the number of seconds to wait for build Bicep files.
  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 10

rule:
  # Enable custom rules that don't exist in the baseline
  includeLocal: true
```

### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.30.0

### Additional context

_No response_

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -28,6 +28,12 @@ See [upgrade notes][1] for helpful information when upgrading from previous vers
 
 ## Unreleased
 
+What's changed since v1.30.0:
+
+- Bug fixes:
+  - Fixed `Azure.Resource.AllowedRegions` which was failing when no allowed regions were configured by @BernieWhite.
+    [#2461](https://github.com/Azure/PSRule.Rules.Azure/issues/2461)
+
 ## v1.30.0
 
 What's changed since v1.29.0:

--- a/src/PSRule.Rules.Azure/Data/Template/ExpressionBuilder.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ExpressionBuilder.cs
@@ -2,10 +2,8 @@
 // Licensed under the MIT License.
 
 using System;
-using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Reflection;
 using System.Threading;
 using Newtonsoft.Json.Linq;
 using PSRule.Rules.Azure.Resources;

--- a/src/PSRule.Rules.Azure/Data/Template/ExpressionHelpers.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ExpressionHelpers.cs
@@ -11,7 +11,6 @@ using System.Security.Cryptography;
 using System.Text;
 using System.Threading;
 using Newtonsoft.Json.Linq;
-using PSRule.Rules.Azure.Resources;
 
 namespace PSRule.Rules.Azure.Data.Template
 {

--- a/src/PSRule.Rules.Azure/Runtime/RuntimeService.cs
+++ b/src/PSRule.Rules.Azure/Runtime/RuntimeService.cs
@@ -74,7 +74,12 @@ namespace PSRule.Rules.Azure.Runtime
             if (locations == null || locations.Length == 0)
                 return;
 
-            _AllowedLocations = new HashSet<string>(locations, LocationHelper.Comparer);
+            _AllowedLocations = new HashSet<string>(LocationHelper.Comparer);
+            for (var i = 0; i < locations.Length; i++)
+            {
+                if (!string.IsNullOrEmpty(locations[i]))
+                    _AllowedLocations.Add(locations[i]);
+            }
         }
 
         /// <inheritdoc/>
@@ -82,6 +87,7 @@ namespace PSRule.Rules.Azure.Runtime
         {
             return location == null ||
                 _AllowedLocations == null ||
+                _AllowedLocations.Count == 0 ||
                 _AllowedLocations.Contains(location);
         }
 

--- a/src/PSRule.Rules.Azure/rules/Conventions.Rule.ps1
+++ b/src/PSRule.Rules.Azure/rules/Conventions.Rule.ps1
@@ -24,7 +24,7 @@ Export-PSRuleConvention 'Azure.Context' -Initialize {
     $minimum = $Configuration.GetValueOrDefault('AZURE_BICEP_MINIMUM_VERSION', '0.4.451');
     $timeout = $Configuration.GetIntegerOrDefault('AZURE_BICEP_FILE_EXPANSION_TIMEOUT', 5);
     $check = $Configuration.GetBoolOrDefault('AZURE_BICEP_CHECK_TOOL', $False);
-    $allowedRegions = @($Configuration.GetValueOrDefault('Azure_AllowedRegions', $Configuration.AZURE_RESOURCE_ALLOWED_LOCATIONS));
+    $allowedRegions = @($Configuration.GetValueOrDefault('Azure_AllowedRegions', $Configuration.GetStringValues('AZURE_RESOURCE_ALLOWED_LOCATIONS')));
     $service = [PSRule.Rules.Azure.Runtime.Helper]::CreateService($minimum, $timeout);
 
     if ($allowedRegions.Length -gt 0) {

--- a/tests/PSRule.Rules.Azure.Tests/RuntimeHelperTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/RuntimeHelperTests.cs
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 using System.Management.Automation;
@@ -51,6 +51,30 @@ namespace PSRule.Rules.Azure
             Assert.True(eval.Outbound("virtualNetwork", 80) == Data.Network.Access.Allow);
         }
 
+        [Fact]
+        public void CreateService()
+        {
+            var service = Helper.CreateService("1.0.0", 90);
+            Assert.NotNull(service);
+            Assert.Equal("1.0.0", service.Minimum);
+            Assert.Equal(90, service.Timeout);
+
+            service.WithAllowedLocations(null);
+            Assert.True(service.IsAllowedLocation("eastus"));
+
+            service.WithAllowedLocations(System.Array.Empty<string>());
+            Assert.True(service.IsAllowedLocation("eastus"));
+
+            service.WithAllowedLocations(new string[] { "westus" });
+            Assert.True(service.IsAllowedLocation("westus"));
+            Assert.False(service.IsAllowedLocation("eastus"));
+
+            service.WithAllowedLocations(new string[] { "" });
+            Assert.True(service.IsAllowedLocation("eastus"));
+        }
+
+        #region Helper methods
+
         private static PSObject NewSecurityRule(int priority, string access, string direction, string destinationAddressPrefix = null, string destinationPortRange = null)
         {
             var result = new PSObject();
@@ -68,5 +92,7 @@ namespace PSRule.Rules.Azure
             result.Properties.Add(new PSNoteProperty("properties", properties));
             return result;
         }
+
+        #endregion Helper methods
     }
 }

